src/averageAges.js

```diff
@@ -1,61 +1,67 @@
- *
- * @param {object[]} people
- * @param {number} century - optional
- *
- * @return {number}
- */
-function calculateMenAverageAge(people, century) {
-  // write code here
-  // learn how to use array methods like .filter .map .some .every .find .reduce
-  // avoid using loop and forEach
-  // replace `if ()` statement with &&, || or ?:
-  // without nesting
+const filterBySex = (people, sex) => {
+  return people.filter(person => person.sex === sex);
+};
+
+const getAverageAge = (people) => {
+  return people
+    .reduce((sumOfAges, person) =>
+      sumOfAges + (person.died - person.born), 0) / people.length;
+};
+
+const filterMothers = (people, women) => {
+  return women
+    .filter(mother => people
+      .map(daughter => daughter.mother).includes(mother.name));
+};
+
+const filterChildren = (people, mothers) => {
+  const children = people.filter(person => person.mother !== null);
+
+  return children
+    .filter(child => mothers
+      .map(mother => mother.name).includes(child.mother));
+};
+
+const parentChildAgeDiff = (parents, children) => {
+  return children.reduce(
+    (totalAge, child) => totalAge + (child.born - parents.filter(
+      parent => parent.name === child.mother)[0].born), 0);
+};
+
+function calculateMenAverageAge(people, century = 0) {
+  let men = filterBySex(people, 'm');
+
+  men = century > 0
+    ? men.filter(man => Math.ceil(man.died / 100) === century)
+    : men;
+
+  return getAverageAge(men);
 }
 
-/**
- * Implement calculateWomenAverageAge function
- *
- * Function returns average ave of women in array. If `withChildren` is
- * specified then function calculates average age only for women with children
- *
- * Hint: To check if a woman has children you should find the other who mention
- * her as mother.
- *
- * @param {object[]} people
- * @param {boolean} withChildren - optional
- *
- * @return {number}
- */
 function calculateWomenAverageAge(people, withChildren) {
-  // write code here
+  let women = filterBySex(people, 'f');
+
+  women = withChildren === true
+    ? filterMothers(people, women)
+    : women;
+
+  return getAverageAge(women);
 }
 
-/**
- * Implement calculateAverageAgeDiff function.
- *
- * The function returns an average age difference between a mother and her
- * child in the array. (A mother's age at child birth)
- *
- * If `onlyWithSon` is specified then function calculates age difference only
- * for mothers who have son.
- *
- * @param {object[]} people
- * @param {boolean} onlyWithSon - optional
- *
- * @return {number}
- */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // write code here
+  const mothers = filterMothers(people, filterBySex(people, 'f'));
+
+  let children = filterChildren(people, mothers);
+
+  children = onlyWithSon === true
+    ? filterBySex(children, 'm')
+    : children;
+
+  const motherChildAgeDiff = parentChildAgeDiff(mothers, children);
+
+  return motherChildAgeDiff / children.length;
```

Comment:
Why children.length? mothers.length seems more logical

Reply:
it does seem more logical. but the thing is that in this function we add the difference of ages of mother and their child / ren. and because one woman can have many children, but one child only one mother, the number of mothers and children is different. and in this case we need to use the number of children.